com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ShaderProgramBuilder.java

```diff
@@ -52,13 +53,33 @@
 import com.dynamo.bob.fs.IResource;
 import com.dynamo.bob.pipeline.ShaderUtil.Common;
 import com.dynamo.bob.pipeline.ShaderUtil.ES2ToES3Converter;
+import com.dynamo.bob.pipeline.ShaderUtil.VariantTextureArrayFallback;
 import com.dynamo.bob.pipeline.ShaderUtil.SPIRVReflector;
 import com.dynamo.bob.util.Exec;
 import com.dynamo.bob.util.Exec.Result;
 import com.dynamo.graphics.proto.Graphics.ShaderDesc;
 import com.google.protobuf.ByteString;
 
 public abstract class ShaderProgramBuilder extends Builder<ShaderPreprocessor> {
```

Comment:
This base class is getting rather complicated and could use more refactoring and simplification. But on the plus side, we now share 100% of the shader generation code between editor and bob so it's still a net gain I think

Reply:
Maybe in the future we will be able to use ANTL4 we already use for lua, to build a syntax tree and symplify everething related to the parsing, regexps and so on.
something like https://github.com/IrisShaders/glsl-transformer
or our own based on GLSL grammar https://github.com/antlr/grammars-v4/tree/master/glsl